### PR TITLE
5.x: adjust stan setup

### PR DESCRIPTION
**Pull request by LordSimal**
Refs: https://github.com/cakephp/cakephp/pull/16935

We already talked about this in chat and @ADmad gave the suggestion to use phive to install the correct versions for stan locally.

I for myself had to install `gnupg` and `gnupg2` to make phive behave via homebrew but that may be just specific to my setup on MacOS (`brew install gnupg gnupg2`)

Regarding the installed versions: Do we want to keep the `.phive/phars.xml` inside the repo or does everyone have to keep track of the used stan versions via the CI config linked in the contributing doc themselves?

### CODE REVIEW

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -76,11 +76,14 @@ Check the [cakephp-codesniffer](https://github.com/cakephp/cakephp-codesniffer)
 repository to set up the CakePHP standard. The [README](https://github.com/cakephp/cakephp-codesniffer/blob/master/README.md) contains installation info
 for the sniff and phpcs.
 
-To run static analysis tools [PHPStan](https://github.com/phpstan/phpstan) and [Psalm](https://github.com/vimeo/psalm) you first have to install the additional packages via:
+To run static analysis tools [PHPStan](https://github.com/phpstan/phpstan) and [Psalm](https://github.com/vimeo/psalm) you first have to install the additional packages via [phive](https://phar.io/).
 
-    composer stan-setup
+    phive install phpstan@1.9
+    phive install psalm@5.4
```
Comment (ADmad) [inline .github/CONTRIBUTING.md]: Updating this version number will be a forgotten chore. psalm releases new minor every few weeks.

Comment (LordSimal) [reply .github/CONTRIBUTING.md]: replaced it with a placeholder

**Comment by dereuromark** [comment]
What's the advantage of this over the previous way?

**Comment by ADmad** [comment]
Is having instructions for running stan tools in the contributing doc necessary? Even today I doubt except for core devs anybody runs them. Others are asked to update the code only after the errors show up on PR.

**Comment by LordSimal** [comment]
> Is having instructions for running stan tools in the contributing doc necessary?

So we want to instruct each new community member on each new PR how to run stan so CI is not broken for them?

**Comment by ADmad** [comment]
> So we want to instruct each new community member on each new PR how to run stan so CI is not broken for them?

Why do we have to provide them instructions to each new member? stan checks already run for the PR and they can see any reported errors on it.

**Comment by ADmad** [comment]
We can keep the existing composer scripts too if someone wants to keep the versions up to date with the ones used in the common Github action.

**Comment by LordSimal** [comment]
> stan checks already run for the PR and they can see any reported errors on it.

sure users can see stan errors in github but from my own experience I can tell, that stan errors are not as 100% clear (especially for new users) and therefore can't be fixed in 1 additional commit if you don't know what you are doing.

So you would say, that users triggering the whole CI suite each time they try to run another "possible stan fix" commit is fine?

what is so wrong documenting and telling users how to run, test and stan our codebase locally?

@dereuromark the old way had the issue, that we have to manually update our composer.json each time we update our `cakephp/.github` stan CI config. This resulted in different stan versions being used locally via `composer stan-setup` then what is run in CI (see #16935)

with pive we remove that burden from us to update the `composer.json` for each repo and instead we/the users have to update our only locally installed phpstan and psalm versions. Therefore we don't "trust" `composer stan-setup` anymore but instead have to keep up2date with what is happening in the CI config.

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -76,11 +76,14 @@ Check the [cakephp-codesniffer](https://github.com/cakephp/cakephp-codesniffer)
 repository to set up the CakePHP standard. The [README](https://github.com/cakephp/cakephp-codesniffer/blob/master/README.md) contains installation info
 for the sniff and phpcs.
 
-To run static analysis tools [PHPStan](https://github.com/phpstan/phpstan) and [Psalm](https://github.com/vimeo/psalm) you first have to install the additional packages via:
+To run static analysis tools [PHPStan](https://github.com/phpstan/phpstan) and [Psalm](https://github.com/vimeo/psalm) you first have to install the additional packages via [phive](https://phar.io/).
 
-    composer stan-setup
+    phive install phpstan@<version>
```
Comment (dereuromark) [inline .github/CONTRIBUTING.md]: Those commands seem rather like a regression
As a user I dont want to know the versions needed, just quickly install and run the existing versions currently locked in IMO.

**Comment by LordSimal** [comment]
Alright, then I'll try to keep our 5.x repos up2date with the CI config.

**Comment by dereuromark** [comment]
You could just commit the xml config file in this repo, no?
It would still be better than having the composer script (with lock hack), it would also be easier/quicker to install all tooling I guess?

**Comment by LordSimal** [comment]
But then its no different than what we already have - having to keep repo stan config in sync with CI stan config.

**Comment by dereuromark** [comment]
Ok, then CI should execute scripts. Fair enough.

**Comment by LordSimal** [comment]
Well (as far as I know) the whole reason to have the CI config files in `cakephp/.github` was the fact, that we didn't want to manage the CI config for each repo we own.

I though we wanted to go that route so there is only 1 place to upgrade stan tool versions instead of having to do it on a per repo base. But if that changed then sure, we can call `composer stan-setup` again in the reusable workflow.

Another thing that came to my mind would be to pass down the stan tool versions to the reusable CI workflow, but that seems to not be easily possible/intended. https://stackoverflow.com/questions/73305126/passing-env-variable-inputs-to-a-reusable-workflow

**Comment by LordSimal** [comment]
Current suggestion from @dereuromark  would be to add the `.phive/phars.xml` to the repo do define the stan versions inside the repo. This will then be executed by CI to install the repo specific versions and stan will be executed via `composer stan` as usual.

With that the `composer stan-setup` hack is removed, CI stan can be adjusted on a per repo basis and everyone using the local stan is on the same version with what is being executed in CI.

**Comment by dereuromark** [comment]
Should be all done now to have it agnostic in execution, while using phinx under the hood.

**Comment by markstory** [comment]
This looks good to me. I'm happy to start using phive if that helps other folks on the team. Would we want to use phive on plugins as well?

**Comment by LordSimal** [comment]
I'd be happy to adjust our plugins cake5 branches as well to be based on this. This is meant to be the start for the new "standard of how to stan our codebase"

**Comment by othercorey** [comment]
Wondering if we should use `.tools` instead of `tools` or do we expect to version control `tools`?

Seems dot-prefixes are no longer temporary directories though with `.github`

**Comment by dereuromark** [comment]
tools/ is not version controlled
Whats the difference? vendor/ is also gitignored and not dot prefixed.
For me that's fine - and easier to type if done manually.
Or do you think there could be a future naming collision of sorts?

**Comment by othercorey** [comment]
Seems unlikely that we will add a tools/ directory and any new framework needs to be more specific than tools.

**Pull request merged**